data_processing_pipeline/6_visualize/src/build_svg.R

```diff
@@ -0,0 +1,127 @@
+
+##### Functions #####
+
+init_svg <- function(width = 8, height = 5, ppi = 72, is_pixels = FALSE) {
+  view_box <- sprintf("%s %s %s %s", 0, 0, ifelse(is_pixels, width, width*ppi), ifelse(is_pixels, height, height*ppi))
+  # create the main "parent" svg node. This is the top-level part of the svg
+  svg_root <- xml_new_root('svg', viewBox = view_box, preserveAspectRatio="xMidYMid meet", 
+                           xmlns="http://www.w3.org/2000/svg", `xmlns:xlink`="http://www.w3.org/1999/xlink", version="1.1" )
+  return(svg_root)
+}
+
+add_state_grp <- function(svg_root, state_nm, trans_x, trans_y) {
+  xml_add_child(svg_root, 'g', id = sprintf('%s-box', state_nm), 
+                transform = sprintf("translate(%s, %s)", trans_x, trans_y))
+}
+
+add_bar_path <- function(svg_root, state_nm, state_data, total_width = 100, 
+                         total_height = 100, round_to = 1) {
+  d <- build_path_from_counts(state_data, total_width = total_width, 
+                              total_height = total_height, round_to = round_to)
+  xml_add_child(svg_root, "path", d = d, id = sprintf('%s-counts', state_nm))
+}
+
+add_hover_rects <- function(svg_root, dat, mx = 0, my = 0, total_width = 100, total_height = 100, round_to = 1) {
+  dat_bars <- format_dat_to_bars(dat, total_width, total_height, round_to) %>% 
+    mutate(x_pos = mx + cumsum(width) - width) # Set up x position to start at mx
+  
+  for(y in dat_bars$year) {
+    dat_y <- filter(dat_bars, year == y) 
+    
+    # add a rectangle for each, add style (don't do it this way in real life) and mouseover events, which won't work 
+    # because hovertext() as a JS function is not defined
+    xml_add_sibling(svg_root, 'rect', x = dat_y$x_pos, y = -total_height, width=dat_y$width, height=total_height, 
+                    style="fill:#0000ff1c",
+                    onmouseover = sprintf("hovertext('%s had %s gages in %s', evt)", dat_y$state, dat_y$n_gages, y),
+                    onmouseout = "hovertext(' ')")
+  }
+  
+  return(svg_root)
+}
+
+build_path_from_counts <- function(dat, mx = 0, my = 0, total_width = 100, 
+                                   total_height = 100, round_to = 1) {
+  
+  dat_bars <- format_dat_to_bars(dat, total_width, total_height, round_to)
+  dat_svg <- dat_bars %>% mutate(v = height - lead(height, 1)) 
+  v_vec <- c(-head(dat_svg$height, 1), # first vertical distance is the first bar's height but negative to move up
+             head(dat_svg$v, -1), # remove the last v bc it will just be NA since there is no `lead` value on the last one
+             tail(dat_svg$height, 1)) # add in height of last bar to get back to baseline, positive to move down 
+  
+  # Build path string with equal widths
+  hv_path_str <- paste(paste0("v", v_vec), collapse = sprintf(" h%s ", unique(dat_bars$width))) # v=vertical, h=horizontal
+  sprintf('M%s,%s %sZ', mx, my, hv_path_str)
+}
+
+format_dat_to_bars <- function(dat, total_width, total_height, round_to) {
+  
```

Comment:
just checking, this is because of the way I have implemented `round_to` where it rounds as it goes rather than at the end, right?

Reply:
No, I think it is actually because of the code below which sets total_width to be a function of how many states you have. I made a comment below when I found the issue